framework/common/message.cpp

```diff
@@ -0,0 +1,21 @@
+/* SPDX-License-Identifier: BSD-2-Clause-Patent
+ *
+ * SPDX-FileCopyrightText: 2016-2020 the prplMesh contributors (see AUTHORS.md)
+ *
+ * This code is subject to the terms of the BSD+Patent license.
+ * See LICENSE file for more details.
+ */
+
+#include <cstdint>
+
+#include <mapf/common/message.h>
+
+namespace mapf {
+
+// Static members definition
+constexpr uint32_t Message::kMessageMagic;
```

Comment:
That's a good observation, and I'm not exactly sure why, but for some reason when there's no declaration for `kMaxFrameLength`, the linker throws the following error:
```
[build:185] - INFO - ['make'] ['map']
[build:70] - INFO - /workspaces/prplMesh/build/. already prepared, skip cmake .
[build:94] - INFO - building & installing .: cmake --build /workspaces/prplMesh/build/. --target install -- -j 7
[23/28] Linking CXX executable out/bin/ieee1905_transport
FAILED: out/bin/ieee1905_transport 
: && /usr/bin/c++  -Wall -Wextra -Werror -pthread -Wno-comment -Wno-unused-parameter -Wno-missing-field-initializers -Wno-address-of-packed-member  -g  -Wl,-rpath=/workspaces/prplMesh/build/install/lib,-rpath-link=/usr/lib    -rdynamic framework/transport/ieee1905_transport/CMakeFiles/ieee1905_transport.dir/main.cpp.o  -o out/bin/ieee1905_transport  -Wl,-rpath,/workspaces/prplMesh/build/out/lib:::: out/lib/libieee1905_transport_lib.so.1.4.0 out/lib/libbtlvf.so.1.4.0 out/lib/libieee1905_transport_messages.so.1.4.0 out/lib/libbcl.so.1.4.0 out/lib/libtlvf.so.1.4.0 /lib/x86_64-linux-gnu/libnl-3.so /usr/lib/x86_64-linux-gnu/libnl-route-3.so out/lib/libmapfcommon.so.1.4.0 out/lib/libelpp.so.1.4.0 -lssl -lcrypto && :
out/lib/libieee1905_transport_messages.so.1.4.0: undefined reference to `mapf::Message::kMaxFrameLength'
collect2: error: ld returned 1 exit status
[25/28] Linking CXX executable out/bin/beerocks_agent
FAILED: out/bin/beerocks_agent 
: && /usr/bin/c++  -Wall -Wextra -Werror -pthread -Wno-comment -Wno-unused-parameter -Wno-missing-field-initializers -Wno-address-of-packed-member  -g  -Wl,-rpath=/workspaces/prplMesh/build/install/lib,-rpath-link=/usr/lib    -rdynamic -Wl,-z,defs agent/src/beerocks/slave/CMakeFiles/beerocks_agent.dir/backhaul_manager/backhaul_manager_thread.cpp.o agent/src/beerocks/slave/CMakeFiles/beerocks_agent.dir/backhaul_manager/wan_monitor.cpp.o agent/src/beerocks/slave/CMakeFiles/beerocks_agent.dir/link_metrics/ieee802_11_link_metrics_collector.cpp.o agent/src/beerocks/slave/CMakeFiles/beerocks_agent.dir/link_metrics/ieee802_3_link_metrics_collector.cpp.o agent/src/beerocks/slave/CMakeFiles/beerocks_agent.dir/platform_manager/platform_manager_thread.cpp.o agent/src/beerocks/slave/CMakeFiles/beerocks_agent.dir/tasks/task_pool.cpp.o agent/src/beerocks/slave/CMakeFiles/beerocks_agent.dir/gate/1905_beacon_query_to_vs.cpp.o agent/src/beerocks/slave/CMakeFiles/beerocks_agent.dir/gate/vs_beacon_response_to_1905.cpp.o agent/src/beerocks/slave/CMakeFiles/beerocks_agent.dir/agent_ucc_listener.cpp.o agent/src/beerocks/slave/CMakeFiles/beerocks_agent.dir/beerocks_slave_main.cpp.o agent/src/beerocks/slave/CMakeFiles/beerocks_agent.dir/son_slave_thread.cpp.o agent/src/beerocks/slave/CMakeFiles/beerocks_agent.dir/tlvf_utils.cpp.o  -o out/bin/beerocks_agent  -Wl,-rpath,/workspaces/prplMesh/build/out/lib:::: -lrt -ldl out/lib/libbtl.so.1.4.0 out/lib/libbtlvf.so.1.4.0 out/lib/libbwl.so.1.4.0 out/lib/libbpl.so.1.4.0 out/lib/libmapfcommon.so.1.4.0 -lssl -lcrypto out/lib/libbcl.so.1.4.0 out/lib/libtlvf.so.1.4.0 /lib/x86_64-linux-gnu/libnl-3.so /usr/lib/x86_64-linux-gnu/libnl-route-3.so out/lib/libelpp.so.1.4.0 -Wl,-rpath-link,/workspaces/prplMesh/build/out/lib && :
/workspaces/prplMesh/build/out/lib/libieee1905_transport_messages.so.1: undefined reference to `mapf::Message::kMaxFrameLength'
collect2: error: ld returned 1 exit status
[26/28] Linking CXX executable out/bin/beerocks_controller
FAILED: out/bin/beerocks_controller 
: && /usr/bin/c++  -Wall -Wextra -Werror -pthread -Wno-comment -Wno-unused-parameter -Wno-missing-field-initializers -Wno-address-of-packed-member  -g  -Wl,-rpath=/workspaces/prplMesh/build/install/lib,-rpath-link=/usr/lib    -rdynamic -Wl,-z,defs controller/src/beerocks/master/CMakeFiles/beerocks_controller.dir/beerocks_master_main.cpp.o controller/src/beerocks/master/CMakeFiles/beerocks_controller.dir/controller_ucc_listener.cpp.o controller/src/beerocks/master/CMakeFiles/beerocks_controller.dir/son_actions.cpp.o controller/src/beerocks/master/CMakeFiles/beerocks_controller.dir/son_management.cpp.o controller/src/beerocks/master/CMakeFiles/beerocks_controller.dir/son_master_thread.cpp.o controller/src/beerocks/master/CMakeFiles/beerocks_controller.dir/tasks/association_handling_task.cpp.o controller/src/beerocks/master/CMakeFiles/beerocks_controller.dir/tasks/bml_task.cpp.o controller/src/beerocks/master/CMakeFiles/beerocks_controller.dir/tasks/channel_selection_task.cpp.o controller/src/beerocks/master/CMakeFiles/beerocks_controller.dir/tasks/client_locating_task.cpp.o controller/src/beerocks/master/CMakeFiles/beerocks_controller.dir/tasks/client_steering_task.cpp.o controller/src/beerocks/master/CMakeFiles/beerocks_controller.dir/tasks/dynamic_channel_selection_task.cpp.o controller/src/beerocks/master/CMakeFiles/beerocks_controller.dir/tasks/ire_network_optimization_task.cpp.o controller/src/beerocks/master/CMakeFiles/beerocks_controller.dir/tasks/load_balancer_task.cpp.o controller/src/beerocks/master/CMakeFiles/beerocks_controller.dir/tasks/network_health_check_task.cpp.o controller/src/beerocks/master/CMakeFiles/beerocks_controller.dir/tasks/optimal_path_task.cpp.o controller/src/beerocks/master/CMakeFiles/beerocks_controller.dir/tasks/statistics_polling_task.cpp.o controller/src/beerocks/master/CMakeFiles/beerocks_controller.dir/tasks/task.cpp.o controller/src/beerocks/master/CMakeFiles/beerocks_controller.dir/tasks/task_pool.cpp.o controller/src/beerocks/master/CMakeFiles/beerocks_controller.dir/db/db.cpp.o controller/src/beerocks/master/CMakeFiles/beerocks_controller.dir/db/db_algo.cpp.o controller/src/beerocks/master/CMakeFiles/beerocks_controller.dir/db/network_map.cpp.o controller/src/beerocks/master/CMakeFiles/beerocks_controller.dir/db/node.cpp.o  -o out/bin/beerocks_controller  -Wl,-rpath,/workspaces/prplMesh/build/out/lib:::: out/lib/libbpl.so.1.4.0 out/lib/libbtl.so.1.4.0 out/lib/libbtlvf.so.1.4.0 out/lib/libbcl.so.1.4.0 out/lib/libtlvf.so.1.4.0 /lib/x86_64-linux-gnu/libnl-3.so /usr/lib/x86_64-linux-gnu/libnl-route-3.so out/lib/libmapfcommon.so.1.4.0 out/lib/libelpp.so.1.4.0 -lssl -lcrypto -Wl,-rpath-link,/workspaces/prplMesh/build/out/lib && :
/workspaces/prplMesh/build/out/lib/libieee1905_transport_messages.so.1: undefined reference to `mapf::Message::kMaxFrameLength'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

What makes it interesting is that both `kMessageMagic` and `kMaxFrameLength` are defined in the same header file, and used in the same source files, but the "undefined reference" is generated only for `kMaxFrameLength`.

Reply:
It's one of the oldest oddities/niggles of C++: You can't put the entire implementation of a (non-template) class in a header file, because then the compiler doesn't know which object file ("compilation unit") to put it in.
So at least part of the implementation of a (non-template) class needs to live in a source file.
In Arnout's example, everything was part of the same .cpp file, so there was no ambiguity, and no error.

IIRC, this is part of the reason why the STL and boost use so many templates: so that they can side-step this issue, and have the entire implementation in header files.